src/Elastic.Apm/Metrics/MetricsProvider/FreeAndTotalMemoryProvider.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using Elastic.Apm.Api;
+using Elastic.Apm.Metrics.Windows;
+
+namespace Elastic.Apm.Metrics.MetricsProvider
+{
+	/// <summary>
+	/// Returns total and free system memory.
+	/// Currently Windows & Linux -only, no macOS support at the moment.
+	/// </summary>
+	internal class FreeAndTotalMemoryProvider : IMetricsProvider
+	{
+		private const string FreeMemory = "system.memory.actual.free";
+		private const string TotalMemory = "system.memory.total";
+
+		public int ConsecutiveNumberOfFailedReads { get; set; }
+		public string NameInLogs => "total and free memory";
+
+		public IEnumerable<Sample> GetValue()
+		{
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				var (success, totalMemory, freeMemory) = GlobalMemoryStatus.GetTotalPhysAndAvailPhys();
+
+				if (!success || totalMemory == 0 || freeMemory == 0)
+					return null;
+
+				return new List<Sample>(2) { new Sample(FreeMemory, freeMemory), new Sample(TotalMemory, totalMemory) };
+			}
+
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+			{
+				var retVal = new List<Sample>();
+
+				using (var sr = new StreamReader("/proc/meminfo"))
+				{
+					var hasMemFree = false;
+					var hasMemTotal = false;
+
+					var line = sr.ReadLine();
+
+					while (line != null || retVal.Count != 2)
+					{
+						if (line != null && line.Contains("MemFree:"))
+						{
+							var (suc, res) = GetEntry(line, "MemFree:");
+							if (suc) retVal.Add(new Sample(FreeMemory, res));
+							hasMemFree = true;
+						}
+						if (line != null && line.Contains("MemTotal:"))
+						{
+							var (suc, res) = GetEntry(line, "MemTotal:");
+							if (suc) retVal.Add(new Sample(TotalMemory, res));
+							hasMemTotal = true;
+						}
+
+						if (hasMemFree && hasMemTotal)
+							break;
+
+						line = sr.ReadLine();
+					}
+				}
+
+				ConsecutiveNumberOfFailedReads = 0;
+
+				return retVal;
+			}
+
+			(bool, ulong) GetEntry(string line, string name)
+			{
+				var nameIndex = line.IndexOf(name, StringComparison.Ordinal);
+				if (nameIndex < 0)
+					return (false, 0);
+
+				var values = line.Substring(line.IndexOf(name, StringComparison.Ordinal) + name.Length);
+
+				if (string.IsNullOrWhiteSpace(values)) return (false, 0);
+
+				var items = values.Trim().Split(' ');
+
+				switch (items.Length)
+				{
+					case 1 when ulong.TryParse(items[0], out var res): return (true, res);
+					case 2 when items[1].ToLower() == "kb" && ulong.TryParse(items[0], out var res): return (true, res * 1024);
+					default: return (false, 0);
+				}
+			}
+
+			return null;
```

Comment:
Addressed in https://github.com/elastic/apm-agent-dotnet/pull/248#discussion_r291309760

Reply:
We are in agreement on not splitting the implementation but then on macOS some metrics providers will be reported as failed in logs even though we know that those metrics providers are not implemented for macOS. I think we should address this issue. One approach to fix would be  to omit not-implemented metrics providers from the list on macOS.

Reply:
This is the current log in macOS:
```
[2019-06-06 05:37:06][Debug] - {MetricsCollector} Metrics collected: MetricSample{system.process.memory.size: 22755934208}, MetricSample{system.process.memory.rss.bytes: 61947904}, MetricSample{system.process.cpu.total.norm.pct: 0.026459536339172}
[2019-06-06 05:37:07][Debug] - {MetricsCollector} Metrics collected: MetricSample{system.process.memory.size: 22755983360}, MetricSample{system.process.memory.rss.bytes: 62988288}, MetricSample{system.cpu.total.norm.pct: 0.0693980497583349}, MetricSample{system.process.cpu.total.norm.pct: 0.00179866871287738}
[2019-06-06 05:37:08][Debug] - {MetricsCollector} Metrics collected: MetricSample{system.process.memory.size: 22755983360}, MetricSample{system.process.memory.rss.bytes: 63893504}, MetricSample{system.cpu.total.norm.pct: 0.0832494607615464}, MetricSample{system.process.cpu.total.norm.pct: 0.00143476148632397}
[2019-06-06 05:37:09][Debug] - {MetricsCollector} Metrics collected: MetricSample{system.process.memory.size: 22755983360}, MetricSample{system.process.memory.rss.bytes: 64774144}, MetricSample{system.cpu.total.norm.pct: 0.0689854589473643}, MetricSample{system.process.cpu.total.norm.pct: 0.00142150237388052}
[2019-06-06 05:37:10][Info] - {MetricsCollector} Failed reading total and free memory 5 consecutively - the agent won't try reading total and free memory anymore
```

>  but then on macOS some metrics providers will be reported as failed in logs 

That's not really the case. We only report explicit failures in `MetricsCollector` when there is an exception - returning null just silently skips and counts the failedReads:

```
try
{
	var samplesFromCurrentProvider = metricsProvider.GetSamples();
	if (samplesFromCurrentProvider != null)
	{
		var sampleArray = samplesFromCurrentProvider as MetricSample[] ?? samplesFromCurrentProvider.ToArray();
		if (sampleArray.Any())
			samplesFromAllProviders.AddRange(sampleArray);
		metricsProvider.ConsecutiveNumberOfFailedReads = 0;
	}
	else
		metricsProvider.ConsecutiveNumberOfFailedReads++;
}
catch (Exception e)
{
	metricsProvider.ConsecutiveNumberOfFailedReads++;
	_logger.Error()
		?.LogException(e, "Failed reading {ProviderName} {NumberOfFail} times", metricsProvider.DbgName,
			metricsProvider.ConsecutiveNumberOfFailedReads);
}
```

So I think the logs are fine, we just print 1 info that this metric is not available. The reason I do this is that there is 1 chart for this on the APM UI, which will be empty, and the fact that it will be the case and an explanation why this is the case I'd say deserves an info log

Reply:
Sounds good. I would have logged the fact that certain metrics providers are not implemented for macOS when not adding them to the list but your way works too. Whichever you prefer.